### ISSUE TITLE
SignalR SourceGenerator attribute API

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/aspnetcore#39973]
## Background and Motivation

For PR https://github.com/dotnet/aspnetcore/pull/38025

We are adding a source generator for SignalR .NET clients to allow sharing strongly-typed interfaces between client and server. We have had the server side since first release and are now adding support for the client-side.

## Proposed API

```diff
namespace Microsoft.AspNetCore.SignalR.Client
{
+    [AttributeUsage(AttributeTargets.Method)]
+    public class ClientHubAttribute : Attribute
+    {
+    }

+    [AttributeUsage(AttributeTargets.Method)]
+    public class ServerHubProxyAttribute : Attribute
+    {
+    }
}
```

Package Name:
Microsoft.AspNetCore.SignalR.Client.SourceGenerator
Attribute DLL Name:
Microsoft.AspNetCore.SignalR.Client.SourceGenerator.HubProxyAttributes.dll

## Usage Examples

```csharp
internal static partial class ProxyExtensions
{
    [ServerHubProxy]
    public static partial T ServerHub<T>(this HubConnection conn);

    [ClientHub]
    public static partial IDisposable CallbackRegistration<T>(this HubConnection conn, T impl);
}

public interface IMyHub
{
    Task GetNothing();
}

public interface IMyHubClient
{
    Task ReceiveMessage(string message);
}

public class MyHubClient : IMyHubClient
{
    public Task ReceiveMessage(string message)
    {
        Console.WriteLine(message);
        return Task.CompletedTask;
    }
}

var myHub = conn.ServerHub<IMyHub>();
var dispose = connection.CallbackRegistration<IMyHubClient>(new MyHubClient());
await myHub.GetNothing();
```

## Alternative Designs

HubServerProxyAttribute

HubClientProxyAttribute
ClientMethodsAttribute
SignalRClientAttribute
ClientRegistrationAttribute

**Comment by pranavkm** [comment]
API review: We were still figuring out the names. The runtime recently added a source generator for P/Invokes and used the name: `GeneratedDllImportAttribute`. It would be nice to try and use that as a naming pattern for similar generated code across frameworks. However we could not come up with a similar name that wasn't also too verbose. For now, the two best options we could come up with were:

```diff
{
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class ClientRegistrationAttribute : Attribute
+    {
+    }

+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class ServerHubProxyAttribute : Attribute
+    {
+    }
```

@BrennanConroy will talk to the remaining SignalR team and determine if these names are good to proceed with.

**Comment by sommmen** [comment]
@BrennanConroy 

Hello, it seems a preview package is on nuget already - i was wondering about the status of client generation for signalR, would the preview package be production ready and will this feature come to net8?

**Comment by BrennanConroy** [comment]
> would the preview package be production ready and will this feature come to net8?

No. See current thoughts on what needs to be done still https://github.com/dotnet/aspnetcore/issues/32534#issuecomment-1668184279
